gunrock/applications/sssp/sssp_implementation.hxx

```diff
@@ -11,165 +11,131 @@
  */
 #pragma once
 
-#include <gunrock/applications/application.hxx>
 #include <bits/stdc++.h>
 
+#include <gunrock/applications/application.hxx>
+
 namespace gunrock {
 namespace sssp {
 
 template <typename meta_t>
-struct sssp_param_t {
-   using vertex_t = typename meta_t::vertex_type;
-   
-   vertex_t single_source;
+struct param_t {
+  using vertex_t = typename meta_t::vertex_type;
    
-   sssp_param_t(
-     vertex_t single_source_
-   ) {
-     single_source = single_source_;
-   }
+  vertex_t single_source;
+  param_t(
+    vertex_t single_source_
+  ) :
+    single_source(single_source_) 
+  {}
 };
 
 template <typename meta_t>
-struct sssp_result_t {
+struct result_t {
   using vertex_t = typename meta_t::vertex_type;
   using weight_t = typename meta_t::weight_type;
    
   weight_t* distances;
   vertex_t* predecessors;
   
-  sssp_result_t(
+  result_t(
     weight_t* distances_,
     vertex_t* predecessors_
-  ) {
-    distances    = distances_;
-    predecessors = predecessors_;
-  }
+  ) :
+    distances(distances_), 
+    predecessors(predecessors_) 
+  {}
 };
 
-
-template <typename d_graph_t, typename meta_t>
-struct sssp_problem_t : problem_t<d_graph_t, meta_t> {
-
-  using param_t   = sssp_param_t<meta_t>;
-  using result_t  = sssp_result_t<meta_t>;
-
+template <typename graph_t, typename meta_t, typename param_t, typename result_t>
+struct problem_t : gunrock::problem_t<graph_t, meta_t, param_t, result_t> {
+  // Use Base class constructor -- does this work? does it handle copy constructor?
+  using gunrock::problem_t<graph_t, meta_t, param_t, result_t>::problem_t;
+  
   using vertex_t = typename meta_t::vertex_type;
   using edge_t   = typename meta_t::edge_type;
   using weight_t = typename meta_t::weight_type;
-
-  vertex_t single_source;
-  weight_t* distances;
-  vertex_t* predecessors;
   
   thrust::device_vector<vertex_t> visited;
-
-  sssp_problem_t(d_graph_t* d_G,
-                 meta_t* meta,
-                 std::shared_ptr<cuda::multi_context_t> context,
-                 param_t& param,
-                 result_t& result)
-      : problem_t<d_graph_t, meta_t>(d_G, meta, context) {
-    
-    single_source = param.single_source;
-    distances     = result.distances;
-    predecessors  = result.predecessors;
-    
-    auto n_vertices = meta[0].get_number_of_vertices();
+  
+  void init() {
+    auto n_vertices = this->get_meta_pointer()->get_number_of_vertices();
+    visited.resize(n_vertices);
+    thrust::fill(thrust::device, visited.begin(), visited.end(), -1);
+  }
+  
+  void reset() {
```

Comment:
I see value in `init()` and `reset()` because separating these allow you to maybe call the algorithm again without the constructor setup, I have, however, not seen this gets used much in the old Gunrock. I found it much cleaner to set-up the problem in the constructor. I don't know if this is the case where we should go with a cleaner code vs. a bit more "programmable".

Reply:
I like the idea of `init` and `reset` because they are logically two different things (stuff you need to do at the beginning, vs stuff you need to do before every invocation).  They also give the user more control over the underlying data.  However, if the data allocation / data initialization in the constructor is fast enough, maybe this doesn't matter very much.

I do also like that this removes a few lines of boilerplate code, passing {G, meta, context, etc) through to the problem constructore.

Reply:
Yeah, I see that. Let's go with your version of using `init` and `reset` for now, we can always move things around later since this isn't crazy tied with anything else.